### PR TITLE
Mcs 861 not going toward target

### PR DESCRIPTION
**Pull request by cedorman**
Added scorecard where the agent needs to go toward the target after having seen the target.  'Seen' means that it was visible for 4 steps (to account for glances and partial observations) and then not decreasing distance from target within 38 steps.  See README for why 38.

### CODE REVIEW

File: tests/test_scorecard.py [hunk 1]
```diff
@@ -1,15 +1,24 @@
 import logging
 import unittest
 
+import numpy as np
+
 import mcs_scene_ingest
-from scorecard.scorecard import Scorecard, find_closest_container
+from scorecard.scorecard import Scorecard
```
[the remark this answers is not in the document]
Comment (cedorman) [reply tests/test_scorecard.py:7]: I'd love to talk to you about how to do this at some point.  I've tried a number of different things, but can't get it to work smoothly.  The result is that I have to do 'from scorecard import *' in the test, which I didn't want to do.  

I'm going to leave as-is for now.

File: scorecard/scorecard.py [hunk 2]
```diff
@@ -366,14 +398,83 @@ def calc_relook(self):
 
         return self.relooks
 
+    def calc_not_moving_toward_object(self):
+        """Calculate number of times that the agent
+        did not move toward the target"""
+
+        self.not_moving_toward_object = 0
+
+        target_id, target_x, target_z = find_target_location(self.scene)
```
[the remark this answers is not in the document]
Comment (cedorman) [reply scorecard/scorecard.py:407]: Ok, we'll have to remember this or tell the dev psychs to ignore the scorecard for those tasks.  

I've added MCS-960 which is to write the target location in the history file and change the scorecard to use it.

File: scorecard/scorecard.py [hunk 3]
```diff
@@ -28,6 +28,17 @@
...
+# Min number of times in a row that the target has to be visible for us
+# to count it as seen by the agent, because the target can be 'visible'
+# but only in the corner for a single frame.
+SEEN_COUNT_MIN = 4
+
+# How many steps required before the agent has to have moved toward the
+# target.  Why 38?  If there is an object between the agent and the target
+# this gives it time to turn (9 steps), move (about 10 steps), turn (9),
+# and move towards it (10).  That's 38 steps.
+STEPS_NOT_MOVED_TOWARD_LIMIT = 38
```
[the remark this answers is not in the document]
Comment (cedorman) [reply scorecard/scorecard.py:40]: In response to this comment, I have change the logic.  It makes sense that the scorecard value is 'Not Moving Toward' that we should only count move* actions.  In particular, if the agent looks up, down, turns, passes, etc, we do not count it.  so, we only count moveahead, back, left, and right.  

If we are only counting real movements, then it makes sense to make the limit the amount it takes to actually move around, which I'm going to say is 15 movements one way (to side) and then 15 towards the target, so 30.

File: scorecard/scorecard.py [hunk 4]
```diff
@@ -366,14 +398,83 @@ def calc_relook(self):
...
+        logging.debug(f"Target location:  {target_x}  {target_z}")
+        if target_id is None:
+            return self.not_moving_toward_object
+
+        seen_count = -1
+        steps_not_moving_towards = 0
+        min_dist = float('inf')
+
+        steps_list = self.history['steps']
+        for step_num, single_step in enumerate(steps_list):
```
[the remark this answers is not in the document]
Comment (cedorman) [reply scorecard/scorecard.py:417]: See below.  Now ignoring all non-Move* actions.

File: tests/test_scorecard.py [hunk 5]
```diff
@@ -1,15 +1,24 @@
...
 TEST_SCENE_FILE_NAME = "occluders_0001_17_I1_debug.json"
 TEST_HISTORY_FILE_NAME = "india_0003_baseline_level1.json"
+
 TEST_SCENE_CONTAINER = "golf_0018_15_debug.json"
+TEST_HISTORY_CONTAINER = "golf_0018_15_baseline.json"
+
+TEST_SCENE_NO_TARGET = "juliett_0001_01_debug.json"
 
 TEST_FOLDER = "tests"
+TEST_FOLDER = '/home/clark/work/mcs/mcs-ingest/tests/'
```
Comment (deanwetherby) [inline tests/test_scorecard.py:21]: Local test folder path should be removed.

File: scorecard/scorecard.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThomasSchellenbergNextCentury) [reply scorecard/scorecard.py]: Sounds great. Thanks!

File: tests/test_scorecard.py (hunk 5 above)
Comment (cedorman) [reply tests/test_scorecard.py]: D'oh.  Removed.

**Pull request merged**